File: internal/tmuxctl/tmux.go
// Package tmuxctl manages tmux sessions for slis slices.
// Each slice gets a session named "slis/<slice>" (sanitised), with one window
// per SliceMember whose cwd is the member's WorktreePath.
package tmuxctl

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/jonnyom/slis/internal/model"
)

var sanitiser = strings.NewReplacer(":", "-", ".", "-")

// SessionName returns the tmux session name for a slice. tmux disallows ':' and
// '.' in session names, so they are replaced with '-'. Format: "slis/<slice>".
func SessionName(slice string) string {
	return "slis/" + sanitiser.Replace(slice)
}

// Available reports whether the tmux binary is on PATH.
func Available() bool {
	_, err := exec.LookPath("tmux")
	return err == nil
}

// SessionExists reports whether the slice's tmux session exists.
func SessionExists(slice string) bool {
	err := exec.Command("tmux", "has-session", "-t", SessionName(slice)).Run()
	return err == nil
}

// SessionOpts controls the window layout of a slice's tmux session.
type SessionOpts struct {
	// Root is the workspace root. When set and the layout includes it, the
	// session opens a window there (so you can run Claude across the whole stack).
	Root string
	// Layout is "root", "repos", or "both". Empty defaults to "root" when Root is
	// set, else "repos".
	Layout string
}

// window is one tmux window: a name and a starting directory.
type window struct{ name, cwd string }

// perRepoWindows builds one window per member, each cd'd into its worktree.
func perRepoWindows(sorted []model.SliceMember) []window {
	wins := make([]window, 0, len(sorted))
	for _, m := range sorted {
		wins = append(wins, window{name: m.Repo, cwd: m.WorktreePath})
	}
	return wins
}

// rootWindowCwd returns the directory a "root" window should cd into so agents
// launched there operate on the slice worktrees, not the repos' primary
// checkouts. For a single member that is the member's worktree; for several it
// is their shared immediate parent (the created-slice layout,
// ws.Root/.slis/worktrees/<slice>/). ok is false when the members do not share
// one immediate parent (adopted/discovered slices with arbitrary paths), in
// which case a single root window cannot serve them all.
func rootWindowCwd(sorted []model.SliceMember) (string, bool) {
	if len(sorted) == 0 {
		return "", false
	}
	if len(sorted) == 1 {
		return sorted[0].WorktreePath, true
	}
	parent := filepath.Dir(sorted[0].WorktreePath)
	for _, m := range sorted[1:] {
		if filepath.Dir(m.WorktreePath) != parent {
			return "", false
		}
	}
	return parent, true
}

// sessionWindows builds the ordered window list for the given members + opts.
func sessionWindows(members []model.SliceMember, opts SessionOpts) []window {
	sorted := make([]model.SliceMember, len(members))
	copy(sorted, members)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Repo < sorted[j].Repo })

	layout := opts.Layout
	if layout == "" {
		if opts.Root != "" {
			layout = "root"
		} else {
			layout = "repos"
		}
	}

	var wins []window
	if (layout == "root" || layout == "both") && opts.Root != "" {
		cwd, ok := rootWindowCwd(sorted)
		if !ok {
			// Members don't share a common parent: a single root window can't
			// serve them all, so fall back to per-repo windows for this session.
			return perRepoWindows(sorted)
		}
		wins = append(wins, window{name: "root", cwd: cwd})
	}
	if layout == "repos" || layout == "both" {
		wins = append(wins, perRepoWindows(sorted)...)
	}
	// Fallback: if the chosen layout produced nothing (e.g. "root" with no Root),
	// fall back to per-repo windows so the session is still useful.
	if len(wins) == 0 {
		return perRepoWindows(sorted)
	}
	return wins
}

// detachHint is the reminder shown in a slis session's status bar. Claude exits
// on Ctrl-D (EOF), which users reach for instinctively when they mean "leave";
// the correct, Claude-preserving way out is the tmux prefix detach (C-b d).
const detachHint = " detach: C-b d  (Ctrl-D quits Claude) "

// setStatusHint writes detachHint into the named session's status bar. It is set
// per-session (no -g) so it only affects slis-owned sessions, never the user's
// other tmux sessions. Best-effort: any failure (e.g. status bar disabled) is
// ignored — it is a hint, not load-bearing.
func setStatusHint(name string) {
	_ = exec.Command("tmux", "set-option", "-t", name, "status-right-length", "40").Run()
	_ = exec.Command("tmux", "set-option", "-t", name, "status-right", detachHint).Run()
}

// EnsureSession creates the slice's tmux session (detached) if it does not
// already exist, with windows determined by opts (a root window, per-repo
// windows, or both — see SessionOpts). Idempotent: returns nil if the session
// already exists. The first window becomes the session's initial (attached) one.
// Either way the session's status bar is (re)stamped with the detach hint.
func EnsureSession(slice string, members []model.SliceMember, opts SessionOpts) error {
	name := SessionName(slice)
	if SessionExists(slice) {
		setStatusHint(name)
		return nil
	}

	wins := sessionWindows(members, opts)
	if len(wins) == 0 {
		// No members and no root: create a bare session.
		if out, err := exec.Command("tmux", "new-session", "-d", "-s", name).CombinedOutput(); err != nil {
			return fmt.Errorf("tmux new-session: %w: %s", err, out)
		}
		setStatusHint(name)
		return nil
	}

	first := wins[0]
	args := []string{"new-session", "-d", "-s", name, "-n", first.name}
	if first.cwd != "" {
		args = append(args, "-c", first.cwd)
	}
	if out, err := exec.Command("tmux", args...).CombinedOutput(); err != nil {
		return fmt.Errorf("tmux new-session: %w: %s", err, out)
	}

	for _, w := range wins[1:] {
		a := []string{"new-window", "-t", name, "-n", w.name}
		if w.cwd != "" {
			a = append(a, "-c", w.cwd)
		}
		if out, err := exec.Command("tmux", a...).CombinedOutput(); err != nil {
			return fmt.Errorf("tmux new-window %q: %w: %s", w.name, err, out)
		}
	}

	setStatusHint(name)
	return nil
}

// CapturePane returns the visible contents of each window's active pane in the
// slice's session, prefixed by a "── <window> ──" header — a read-only peek at
// what is running (e.g. a Claude session) without attaching. Returns an error
// if the session does not exist.
func CapturePane(slice string) (string, error) {
	name := SessionName(slice)
	out, err := exec.Command("tmux", "list-windows", "-t", name, "-F", "#{window_index}\t#{window_name}").Output()
	if err != nil {
		return "", fmt.Errorf("tmux list-windows: %w", err)
	}

	var sb strings.Builder
	for _, line := range strings.Split(strings.TrimRight(string(out), "\n"), "\n") {
		if line == "" {
			continue
		}
		idx, wname, _ := strings.Cut(line, "\t")
		// -e preserves the pane's colour escapes (slis keeps SGR, strips the rest).
		captured, _ := exec.Command("tmux", "capture-pane", "-p", "-e", "-t", name+":"+idx).Output()
		sb.WriteString("── " + wname + " ──\n")
		sb.Write(captured)
		if !strings.HasSuffix(string(captured), "\n") {
			sb.WriteString("\n")
		}
		sb.WriteString("\n")
	}
	return sb.String(), nil
}

// SendKeys types keys into the slice session's active pane followed by Enter
// (used to launch an agent). No-op error if the session is absent.
func SendKeys(slice, keys string) error {
	err := exec.Command("tmux", "send-keys", "-t", SessionName(slice), keys, "Enter").Run()
	if err != nil {
		return fmt.Errorf("tmux send-keys: %w", err)
	}
	return nil
}

// ActivePaneCommand returns the foreground command of the session's active pane
// (e.g. "zsh", "node", "claude"), or "" if it can't be determined.
func ActivePaneCommand(slice string) string {
	out, err := exec.Command("tmux", "display-message", "-p", "-t", SessionName(slice), "#{pane_current_command}").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// PanePIDs returns the pane PIDs across all windows of the slice's session.
func PanePIDs(slice string) ([]int, error) {
	name := SessionName(slice)
	out, err := exec.Command("tmux", "list-panes", "-s", "-t", name, "-F", "#{pane_pid}").Output()
	if err != nil {
		return nil, fmt.Errorf("tmux list-panes: %w", err)
	}

	var pids []int
	for _, line := range strings.Split(string(out), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		pid, err := strconv.Atoi(line)
		if err != nil {
			return nil, fmt.Errorf("parsing pane pid %q: %w", line, err)
		}
		pids = append(pids, pid)
	}
	return pids, nil
}

// AttachArgv returns the argv to attach to the slice's session, choosing
// switch-client when already inside tmux ($TMUX set) or attach otherwise.
// Returned as (name, args) so it is unit-testable without spawning tmux.
func AttachArgv(slice string, insideTmux bool) (string, []string) {
	target := SessionName(slice)
	if insideTmux {
		return "tmux", []string{"switch-client", "-t", target}
	}
	return "tmux", []string{"attach", "-t", target}
}

// Attach attaches the current terminal to the slice's session (switch-client if
// inside tmux, else attach with inherited stdio). Blocks until detach when attaching.
func Attach(slice string) error {
	name, args := AttachArgv(slice, os.Getenv("TMUX") != "")
	cmd := exec.Command(name, args...)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

// Client is a single attached tmux client: its controlling TTY, the session it
// is currently viewing, and its last-activity time (unix seconds) used to pick
// the most recently used client to steal focus onto.
type Client struct {
	TTY          string
	Session      string
	LastActivity int64
}

// clientListFormat is the -F format string for ListClients: tty, session name,
// and last-activity timestamp, tab-separated.
const clientListFormat = "#{client_tty}\t#{client_session}\t#{client_activity}"

// parseClients parses `tmux list-clients -F clientListFormat` output. Lines that
// are blank, malformed, or missing a TTY are skipped. Pure — unit-testable
// without spawning tmux.
func parseClients(raw string) []Client {
	var clients []Client
	for _, line := range strings.Split(strings.TrimRight(raw, "\n"), "\n") {
		if line == "" {
			continue
		}
		parts := strings.Split(line, "\t")
		if len(parts) < 3 {
			continue
		}
		tty := strings.TrimSpace(parts[0])
		if tty == "" {
			continue
		}
		activity, _ := strconv.ParseInt(strings.TrimSpace(parts[2]), 10, 64)
		clients = append(clients, Client{
			TTY:          tty,
			Session:      strings.TrimSpace(parts[1]),
			LastActivity: activity,
		})
	}
	return clients
}

// ListClients returns the currently attached tmux clients. tmux only lists
// attached clients, so an empty slice means nobody is attached anywhere.
func ListClients() ([]Client, error) {
	out, err := exec.Command("tmux", "list-clients", "-F", clientListFormat).Output()
	if err != nil {
		return nil, fmt.Errorf("tmux list-clients: %w", err)
	}
	return parseClients(string(out)), nil
}

// MostRecentClient returns the client with the greatest last-activity time (the
// one the user most recently used). ok is false when clients is empty.
func MostRecentClient(clients []Client) (Client, bool) {
	if len(clients) == 0 {
		return Client{}, false
	}
	best := clients[0]
	for _, c := range clients[1:] {
		if c.LastActivity > best.LastActivity {
			best = c
		}
	}
	return best, true
}

// SwitchClientArgv returns the argv to point a specific client (by its TTY) at
// the slice's session. Returned as (name, args) so it is unit-testable without
// spawning tmux.
func SwitchClientArgv(clientTTY, slice string) (string, []string) {
	return "tmux", []string{"switch-client", "-c", clientTTY, "-t", SessionName(slice)}
}

// KillSession kills the slice's session (used for cleanup/teardown).
// Returns nil if the session does not exist.
func KillSession(slice string) error {
	err := exec.Command("tmux", "kill-session", "-t", SessionName(slice)).Run()
	if err != nil {
		// Tolerate "session not found" errors.
		if !SessionExists(slice) {
			return nil
		}
		return fmt.Errorf("tmux kill-session: %w", err)
	}
	return nil
}

File: tui-js/src/term/tmux.ts
// tmux + agent-launch helpers, ported from the Go TUI so the JS terminal tabs
// create and populate a slice's session identically:
//   - session naming            → internal/tmuxctl.SessionName
//   - EnsureSession windows      → internal/tmuxctl.EnsureSession / sessionWindows
//   - SLIS_* env + slice context → internal/tui.agentLaunchLine (agentctx.go)
//
// We never own the agent process — we `tmux attach` a client (see session.ts).
// Everything here is a thin shell-out to the `tmux` binary; nothing mutates a repo.

import { dirname, isAbsolute, relative, resolve } from "node:path";

/** A slice member reduced to what session windows + agent context need. */
export interface TermMember {
  repo: string;
  branch: string;
  worktreePath: string;
}

/** Window-layout options, mirroring internal/tmuxctl.SessionOpts. */
export interface SessionOpts {
  /** Workspace root; enables a "root"/"both" window at the slice's shared parent. */
  root?: string;
  /** "root" | "repos" | "both". Empty → root when members share a parent. */
  layout?: string;
}

/** Agent and ad-hoc shell terminals intentionally use separate tmux sessions. */
export type SessionKind = "agent" | "shell";

export interface TmuxPane {
  path: string;
  command: string;
  target?: string;
}

export interface TmuxSessionInfo {
  name: string;
  kind: SessionKind;
  panes: TmuxPane[];
}

/** tmux disallows ':' and '.' in session names → replace with '-'. */
export function sessionName(slice: string, kind: SessionKind = "agent"): string {
  const prefix = kind === "agent" ? "slis/" : "slis-shell/";
  return prefix + slice.replace(/[:.]/g, "-");
}

async function sh(cmd: string[]): Promise<{ code: number; out: string }> {
  const p = Bun.spawn(cmd, { stdout: "pipe", stderr: "pipe", stdin: "ignore" });
  const out = await new Response(p.stdout).text();
  const code = await p.exited;
  return { code, out };
}

export function tmuxAvailable(): boolean {
  return Bun.which("tmux") !== null;
}

export function parseTmuxSessions(output: string): TmuxSessionInfo[] {
  const sessions = new Map<string, TmuxSessionInfo>();
  for (const line of output.split("\n")) {
    if (!line) continue;
    const [name, path, command, target] = line.split("\t");
    if (!name || !path || !command) continue;
    const kind = name.startsWith("slis/")
      ? "agent"
      : name.startsWith("slis-shell/")
        ? "shell"
        : null;
    if (!kind) continue;
    const session = sessions.get(name) ?? { name, kind, panes: [] };
    session.panes.push({ path, command, target });
    sessions.set(name, session);
  }
  return [...sessions.values()].sort((a, b) => a.name.localeCompare(b.name));
}

export async function listTmuxSessions(): Promise<TmuxSessionInfo[]> {
  const result = await sh([
    "tmux",
    "list-panes",
    "-a",
    "-F",
    "#{session_name}\t#{pane_current_path}\t#{pane_current_command}\t#{session_name}:#{window_index}.#{pane_index}",
  ]);
  return result.code === 0 ? parseTmuxSessions(result.out) : [];
}

export async function sessionExists(slice: string, kind: SessionKind = "agent"): Promise<boolean> {
  return (await sh(["tmux", "has-session", "-t", sessionName(slice, kind)])).code === 0;
}

/** Current directories for every pane in a slice session. */
export async function sessionPanePaths(slice: string, kind: SessionKind = "agent"): Promise<string[]> {
  const r = await sh([
    "tmux",
    "list-panes",
    "-s",
    "-t",
    sessionName(slice, kind),
    "-F",
    "#{pane_current_path}",
  ]);
  if (r.code !== 0) return [];
  return [...new Set(r.out.split("\n").map((path) => path.trim()).filter(Boolean))];
}

function pathIsWithin(path: string, parent: string): boolean {
  const rel = relative(resolve(parent), resolve(path));
  return rel === "" || (!rel.startsWith("..") && !isAbsolute(rel));
}

/** True when any session pane is operating outside all configured worktrees. */
export function sessionHasPaneOutsideMembers(paths: string[], members: TermMember[]): boolean {
  return paths.some((path) => !members.some((member) => pathIsWithin(path, member.worktreePath)));
}

export function tmuxSessionRelatedToMembers(
  session: TmuxSessionInfo,
  members: TermMember[],
): boolean {
  return session.panes.some((pane) =>
    members.some((member) => pathIsWithin(pane.path, member.worktreePath)),
  );
}

export function preferredRunningAgentSession(
  sessions: TmuxSessionInfo[],
  members: TermMember[],
): TmuxSessionInfo | undefined {
  return sessions.find(
    (session) =>
      session.kind === "agent" &&
      tmuxSessionRelatedToMembers(session, members) &&
      session.panes.some((pane) => !isShellCmd(pane.command)),
  );
}

export async function killTmuxSession(name: string): Promise<boolean> {
  if (!name.startsWith("slis/") && !name.startsWith("slis-shell/")) return false;
  return (await sh(["tmux", "kill-session", "-t", name])).code === 0;
}

export async function resumeClaudeSession(opts: {
  slice: string;
  sessionId: string;
  cwd?: string;
  members: TermMember[];
  sessionOpts: SessionOpts;
}): Promise<string> {
  if (!/^[A-Za-z0-9-]+$/.test(opts.sessionId)) throw new Error("invalid Claude session id");
  await ensureSession(opts.slice, opts.members, opts.sessionOpts, "agent");
  const name = sessionName(opts.slice);
  const session = (await listTmuxSessions()).find((candidate) => candidate.name === name);
	const target = session?.panes.find((pane) => isShellCmd(pane.command))?.target ?? name;
	const root = rootWindowCwd(opts.members);
	const resume = `claude --resume ${opts.sessionId}`;
	const command = root.ok ? `cd ${shellSingleQuote(root.cwd)} && ${resume}` : resume;
  const result = await sh([
    "tmux",
    "send-keys",
    "-t",
    target,
		command,
    "Enter",
  ]);
  if (result.code !== 0) throw new Error(`tmux send-keys: ${result.out}`);
  return name;
}

interface Window {
  name: string;
  cwd: string;
}

function perRepoWindows(sorted: TermMember[]): Window[] {
  return sorted.map((m) => ({ name: m.repo, cwd: m.worktreePath }));
}

// rootWindowCwd returns the directory a single "root" window cd's into so agents
// operate on the slice worktrees. For one member that is its worktree; for many
// it is their shared immediate parent. ok=false when they don't share one.
function rootWindowCwd(sorted: TermMember[]): { cwd: string; ok: boolean } {
  if (sorted.length === 0) return { cwd: "", ok: false };
  if (sorted.length === 1) return { cwd: sorted[0]!.worktreePath, ok: true };
  const parent = dirname(sorted[0]!.worktreePath);
  for (const m of sorted.slice(1)) {
    if (dirname(m.worktreePath) !== parent) return { cwd: "", ok: false };
  }
  return { cwd: parent, ok: true };
}

export function sessionWindows(members: TermMember[], opts: SessionOpts): Window[] {
  const sorted = [...members].sort((a, b) => a.repo.localeCompare(b.repo));

  let layout = opts.layout ?? "";
  if (layout === "") layout = opts.root ? "root" : "repos";

  let wins: Window[] = [];
  if ((layout === "root" || layout === "both") && opts.root) {
    const { cwd, ok } = rootWindowCwd(sorted);
    if (!ok) return perRepoWindows(sorted); // no shared parent → per-repo
    wins.push({ name: "root", cwd });
  }
  if (layout === "repos" || layout === "both") {
    wins = wins.concat(perRepoWindows(sorted));
  }
  if (wins.length === 0) return perRepoWindows(sorted);
  return wins;
}

// Claude exits on Ctrl-D (EOF); the correct, Claude-preserving way out is the
// tmux prefix detach (C-b d). Mirrors internal/tmuxctl.detachHint.
const DETACH_HINT = " detach: C-b d  (Ctrl-D quits Claude) ";

async function setStatusHint(name: string, kind: SessionKind): Promise<void> {
  // Per-session mouse mode lets the embedded client use wheel scrolling without
  // changing the user's global tmux configuration.
  await sh(["tmux", "set-option", "-t", name, "mouse", "on"]);
  await sh(["tmux", "set-option", "-t", name, "status-right-length", "40"]);
  const hint = kind === "agent" ? DETACH_HINT : " detach: C-b d  (ctrl+q returns to Slis) ";
  await sh(["tmux", "set-option", "-t", name, "status-right", hint]);
}

/**
 * Create the slice's tmux session (detached) if it does not already exist, with
 * windows determined by opts. Idempotent. Mirrors tmuxctl.EnsureSession.
 */
export async function ensureSession(
  slice: string,
  members: TermMember[],
  opts: SessionOpts,
  kind: SessionKind = "agent",
): Promise<void> {
  const name = sessionName(slice, kind);
  if (await sessionExists(slice, kind)) {
    await setStatusHint(name, kind);
    return;
  }

  const wins = sessionWindows(members, opts);
  if (wins.length === 0) {
    const r = await sh(["tmux", "new-session", "-d", "-s", name]);
    if (r.code !== 0) throw new Error(`tmux new-session: ${r.out}`);
    await setStatusHint(name, kind);
    return;
  }

  const first = wins[0]!;
  const args = ["new-session", "-d", "-s", name, "-n", first.name];
  if (first.cwd) args.push("-c", first.cwd);
  const created = await sh(["tmux", ...args]);
  if (created.code !== 0) throw new Error(`tmux new-session: ${created.out}`);

  for (const w of wins.slice(1)) {
    const a = ["new-window", "-t", name, "-n", w.name];
    if (w.cwd) a.push("-c", w.cwd);
    const r = await sh(["tmux", ...a]);
    if (r.code !== 0) throw new Error(`tmux new-window ${w.name}: ${r.out}`);
  }

  await setStatusHint(name, kind);
}

/** Foreground command of the session's active pane (e.g. "zsh", "claude"). */
export async function activePaneCommand(slice: string, kind: SessionKind = "agent"): Promise<string> {
  const r = await sh([
    "tmux",
    "display-message",
    "-p",
    "-t",
    sessionName(slice, kind),
    "#{pane_current_command}",
  ]);
  return r.code === 0 ? r.out.trim() : "";
}

/** Whether cmd is an interactive shell (safe to type a launch line into). */
export function isShellCmd(cmd: string): boolean {
  return ["zsh", "bash", "fish", "sh", "dash", "ksh", "tcsh"].includes(cmd);
}

/** Type keys into the session's active pane followed by Enter. */
export async function sendKeys(slice: string, keys: string, kind: SessionKind = "agent"): Promise<void> {
  await sh(["tmux", "send-keys", "-t", sessionName(slice, kind), keys, "Enter"]);
}

// ── agent launch line (ported from internal/tui/agentctx.go) ─────────────────

function isClaudeAgent(agent: string): boolean {
  const bin = agent.trim().split(/\s+/)[0] ?? "";
  return bin === "claude" || bin.endsWith("/claude");
}

function shellSingleQuote(s: string): string {
  return "'" + s.replaceAll("'", `'\\''`) + "'";
}

function slisAgentContext(slice: string, members: TermMember[], active: boolean): string {
  const sorted = [...members].sort((a, b) => a.repo.localeCompare(b.repo));
  const parts = sorted.map((m) =>
    m.worktreePath
      ? `${m.repo} → ${m.worktreePath} (branch ${m.branch})`
      : `${m.repo} (branch ${m.branch})`,
  );
  let ctx =
    `You are running inside slis, a multi-repo worktree cockpit, working on slice "${slice}" ` +
    `which spans ${sorted.length} repo(s). Make ALL your edits inside this slice's git worktrees, listed here — ` +
    `do NOT touch the repos' primary checkouts: ${parts.join("; ")}. Each repo is a separate worktree on its own ` +
    `branch; cd into the right worktree for each repo and keep every commit scoped to that worktree.`;
  if (active) {
    ctx +=
      " (This slice is also LIVE — swapped into the primary checkouts so dev servers build it — " +
      "but still make every edit in the worktrees above, never the primaries.)";
  }
  return ctx;
}

function withSlisContext(agent: string, slice: string, members: TermMember[], active: boolean): string {
  if (!isClaudeAgent(agent)) return agent;
  return agent + " --append-system-prompt " + shellSingleQuote(slisAgentContext(slice, members, active));
}

function slisEnvPrefix(
  slice: string,
  members: TermMember[],
  active: boolean,
  wsRoot: string,
  harness: string,
): string {
  const sorted = [...members].sort((a, b) => a.repo.localeCompare(b.repo));
  const pairs = sorted.map((m) => `${m.repo}=${m.worktreePath}`);
  const vars = [
    "SLIS_SLICE=" + shellSingleQuote(slice),
    "SLIS_ROOT=" + shellSingleQuote(wsRoot),
    "SLIS_ACTIVE=" + shellSingleQuote(active ? "1" : "0"),
    "SLIS_HARNESS=" + shellSingleQuote(harness),
    "SLIS_WORKTREES=" + shellSingleQuote(pairs.join(",")),
  ];
  const terminalApp = process.env.SLIS_TERMINAL_APP ||
    (process.env.TERM_PROGRAM?.toLowerCase() === "ghostty" ? "ghostty" : "");
  if (terminalApp) vars.push("SLIS_TERMINAL_APP=" + shellSingleQuote(terminalApp));
  return vars.join(" ");
}

/**
 * The full one-line launch command: the SLIS_* env prefix followed by the agent
 * command (with claude's --append-system-prompt appended). Mirrors
 * internal/tui.agentLaunchLine.
 */
export function agentLaunchLine(opts: {
  agent: string;
  harness: string;
  slice: string;
  members: TermMember[];
  active: boolean;
  wsRoot: string;
}): string {
  const { agent, harness, slice, members, active, wsRoot } = opts;
	const launch = (
    slisEnvPrefix(slice, members, active, wsRoot, harness) +
    " " +
    withSlisContext(agent, slice, members, active)
  );
	const root = rootWindowCwd(members);
	return root.ok ? `cd ${shellSingleQuote(root.cwd)} && ${launch}` : launch;
}
